subprojects/core/src/integTest/groovy/org/gradle/api/tasks/NestedInputIntegrationTest.groovy

```diff
@@ -655,6 +658,175 @@ class NestedInputIntegrationTest extends AbstractIntegrationSpec {
         fails "myTask"
     }
 
+    def "implementation of nested property in Groovy build script is tracked"() {
+        setupTaskClassWithNestedAction()
+        buildFile << """
+            task myTask(type: TaskWithNestedAction) {
+                action = ${originalImplementation}
+            }
+        """
+
+        buildFile.makeOlder()
+
+        when:
+        run 'myTask'
+        then:
+        executedAndNotSkipped(':myTask')
+
+        when:
+        buildFile.text = """
+            task myTask(type: TaskWithNestedAction) {
+                action = ${changedImplementation}
+            }
+        """
+        run 'myTask', '--info'
+        then:
+        executedAndNotSkipped(':myTask')
+        file('build/tmp/myTask/output.txt').text == "changed"
+        output.contains "Value of input property 'action.class' has changed for task ':myTask'"
+
+        where:
+        originalImplementation      | changedImplementation
+        """{ it.text = "hello" }""" | """{ it.text = "changed" }"""
+        """
+            new Action() {
+                void execute(outputFile) {
+                    outputFile.text = "hello"
+                }
+            }"""                    | """
+                                        new Action() {
+                                            void execute(outputFile) {
+                                                outputFile.text = "changed"
+                                            }
+                                        }"""
```

Comment:
Maybe add something like a `wrapAction()` method, so this can become:

```groovy
where:
originalImplementation                  | changedImplementation
// ...
wrapAction('outputFile.text = "hello"') | wrapAction('outputFile.text = "changed"') 
```